File: src/renderer/menu.c
#include "menu.h"

#include "render.h"
#include "text.h"
#include <stdlib.h>

int drawControl(Control *control, RenderInfo *renderer, Vector2 offset) {
  return (control->draw)(control->control, renderer, offset);
}

int updateControl(Control *control, RenderInfo *renderer) {
  return (control->update)(control->control, renderer);
}


Menu * createMenu() {
  Menu *menu = malloc(sizeof(Menu));

  menu->controls = NULL;
  menu->orientation = Vertical;
  menu->position = vec2(0.0, 0.0);
  menu->size = vec2(0.0, 0.0);
  menu->color = vec4(0.0, 0.0, 0.0, 0.0);
  menu->z_index = 0;
  menu->controls_n = 0;

  return menu;
}

void addControlToMenu(Menu *menu, Control *control) {
  menu->controls = realloc(menu->controls, sizeof(Control *) * (menu->controls_n + 1));
  menu->controls[menu->controls_n] = control;
  ++menu->controls_n;
}

void setMenuOrientation(Menu *menu, MenuOrientation orientation) {
  menu->orientation = orientation;
}

void setMenuPosition(Menu *menu, Vector2 position) {
  menu->position = position;
}

void setMenuSize(Menu *menu, Vector2 size) {
  menu->size = size;
}

void setMenuColor(Menu *menu, Vector4 color) {
  menu->color = color;
}

int drawMenu(Menu *menu, RenderInfo *renderer) {
  // Draw the background
  renderQuad(renderer, menu->size, menu->position, menu->color, menu->z_index / Z_INDEX_MAX);

  for (int i = 0; i < menu->controls_n; ++i) {
    drawControl(menu->controls[i], renderer, menu->position);
  }
  return 0;
}

int updateMenu(Menu *menu, RenderInfo *renderer) {
  for (int i = 0; i < menu->controls_n; ++i) {
    updateControl(menu->controls[i], renderer);
  }
  return 0;
}

Control * createButton() {
  Button *b = malloc(sizeof(Button));

  b->state = Normal;

  Control *c = malloc(sizeof(Control));
  c->control = b;
  c->draw = drawButton;
  c->update = updateButton;

  return c;
}

int drawButton(void *c, RenderInfo *renderer, Vector2 offset) {
  Button *button = (Button *)c;

  Vector4 color;
  if (button->state == Normal)
    color = button->color;
  else if (button->state == Highlighted)
    color = button->highlight;
  else if (button->state == Selected)
    color = button->select;

  renderQuad(renderer, button->size, button->position, color, button->z_index / Z_INDEX_MAX);
  drawText(renderer, button->text, button->position, button->size, (float)button->textHeight / (float)FONT_SIZE, button->textColor);

  return 0;
}

int updateButton(void *c, RenderInfo *renderer) {
  Button *button = (Button *)c;

  Vector2 cursorPosition = getCursorPosition(renderer);
  int leftButton = isLeftMouseButtonPressed(renderer);

  // Test for mouse position
  if (cursorPosition.x >= button->position.x && cursorPosition.y >= button->position.y &&
      cursorPosition.x <= button->position.x + button->size.x && cursorPosition.y <= button->position.y + button->size.y) {
    if (leftButton == 1) {
      // Call the action if the state changed to selected
      if (button->state != Selected && button->action != NULL)
        button->action(button);
      button->state = Selected;
    } else
      button->state = Highlighted;
  } else {
    button->state = Normal;
  }

  return 0;
}

Button * getButton(Control *button) {
  return (Button *)button->control;
}

void setButtonZ(Control *button, unsigned int z) {
  getButton(button)->z_index = z;
}

void setButtonPosition(Control *button, Vector2 position) {
  getButton(button)->position = position;
}

void setButtonSize(Control *button, Vector2 size) {
  getButton(button)->size = size;
}

void setButtonColor(Control *button, Vector4 color) {
  getButton(button)->color = color;
}

void setButtonText(Control *button, char * text) {
  getButton(button)->text = text;
}

void setButtonTextHeight(Control *button, int textHeight) {
  getButton(button)->textHeight = textHeight;
}

void setButtonTextColor(Control *button, Vector4 color) {
  getButton(button)->textColor = color;
}

void setButtonAction(Control *button, void (* action)(struct Button *)) {
  getButton(button)->action = action;
}


Control * createLabel() {
  Label *l = malloc(sizeof(Label));

  Control *c = malloc(sizeof(Control));
  c->control = l;
  c->draw = drawLabel;
  c->update = updateLabel;

  return c;
}

int drawLabel(void *c, RenderInfo *renderer, Vector2 offset) {
  Label *label = (Label *)c;

  Vector2 windowSize = getWindowSize(renderer);

  // Convert pixel coordinates to NDC
  // Position specifies bottom left coordinates
  drawText(renderer, label->text, add2(label->position, offset), label->size, (float)label->textHeight / (float)FONT_SIZE, label->color);

  return 0;
}

int updateLabel(void *c, RenderInfo *renderer) {
  Label *label = (Label *)c;
  return 0;
}

Label * getLabel(Control *label) {
  return (Label *)label->control;
}

void setLabelZ(Control *label, unsigned int z) {
  getLabel(label)->z_index = z;
}

void setLabelPosition(Control *label, Vector2 position) {
  getLabel(label)->position = position;
}

void setLabelSize(Control *label, Vector2 size) {
  getLabel(label)->size = size;
}

void setLabelColor(Control *label, Vector4 color) {
  getLabel(label)->color = color;
}

void setLabelText(Control *label, char * text) {
  getLabel(label)->text = text;
}

void setLabelTextHeight(Control *label, int textHeight) {
  getLabel(label)->textHeight = textHeight;
}

File: src/menu.jl
const Menu = Ptr{Cvoid}
const Control = Ptr{Cvoid}
const Label = Ptr{Cvoid}
const Button = Ptr{Cvoid}

@enum MenuOrientation begin
  Vertical = 0
  Horizontal = 1
  Central = 2
end

# Menu
function createMenu()
  ccall((:createMenu, @fullLibraryPath), Menu, ())
end

function setMenuOrientation(menu, orientation)
  ccall((:setMenuOrientation, @fullLibraryPath), Cvoid, (Menu, MenuOrientation), menu, orientation)
end

function setMenuPosition(menu, position)
  ccall((:setMenuPosition, @fullLibraryPath), Cvoid, (Menu, Vector2), menu, position)
end

function setMenuSize(menu, size)
  ccall((:setMenuSize, @fullLibraryPath), Cvoid, (Menu, Vector2), menu, size)
end

function setMenuColor(menu, color)
  ccall((:setMenuColor, @fullLibraryPath), Cvoid, (Menu, Vector4), menu, color)
end

function addControlToMenu(menu, control)
  ccall((:addControlToMenu, @fullLibraryPath), Cvoid, (Menu, Control), menu, control)
end

function updateMenu(menu)
  ccall((:updateMenu, @fullLibraryPath), Cvoid, (Menu,), menu)
end

# Label
function createLabel()
  ccall((:createLabel, @fullLibraryPath), Label, ())
end

function setLabelZ(label, z)
  ccall((:setLabelZ, @fullLibraryPath), Cvoid, (Label, Cint), label, z)
end

function setLabelPosition(label, position)
  ccall((:setLabelPosition, @fullLibraryPath), Cvoid, (Label, Vector2), label, position)
end

function setLabelSize(label, size)
  ccall((:setLabelSize, @fullLibraryPath), Cvoid, (Label, Vector2), label, size)
end

function setLabelColor(label, color)
  ccall((:setLabelColor, @fullLibraryPath), Cvoid, (Label, Vector4), label, color)
end

function setLabelTextHeight(label, textHeight)
  ccall((:setLabelTextHeight, @fullLibraryPath), Cvoid, (Label, Cint), label, textHeight)
end

function setLabelText(label, text)
  ccall((:setLabelText, @fullLibraryPath), Cvoid, (Label, Cstring), label, text)
end

# Button
function createButton()
  ccall((:createButton, @fullLibraryPath), Button, ())
end

function setButtonZ(button, z)
  ccall((:setButtonZ, @fullLibraryPath), Cvoid, (Button, Cint), button, z)
end

function setButtonPosition(button, position)
  ccall((:setButtonPosition, @fullLibraryPath), Cvoid, (Button, Vector2), button, position)
end

function setButtonSize(button, size)
  ccall((:setButtonSize, @fullLibraryPath), Cvoid, (Button, Vector2), button, size)
end

function setButtonTextHeight(button, textHeight)
  ccall((:setButtonTextHeight, @fullLibraryPath), Cvoid, (Button, Cint), button, textHeight)
end

function setButtonText(button, text)
  ccall((:setButtonText, @fullLibraryPath), Cvoid, (Button, Cstring), button, text)
end

function setButtonAction(button, action)
  ccall((:setButtonAction, @fullLibraryPath), Cvoid, (Button, Ptr{Cvoid}), button, @cfunction($action, Cvoid, (Button,)))
end

function setButtonTextColor(button, textColor)
  ccall((:setButtonTextColor, @fullLibraryPath), Cvoid, (Button, Vector4), button, textColor)
end
